liquibase-core/src/main/java/liquibase/resource/ResourceAccessor.java

```diff
@@ -112,6 +112,30 @@ default SortedSet<String> list(String relativeTo, String path, boolean recursive
         return returnList;
     }
 
+    /**
+     * Returns the path to all resources contained in the given path.
+     * Multiple resources may be returned with the same path, but only if they are actually unique files.
+     * Order is important to pay attention to, they should be returned in a user-expected manner based on this resource accessor.
+     * <br><br>
+     * Should return an empty list if:
+     * <ul>
+     *     <li>Path does not exist</li>
+     * </ul>
+     * Should throw an exception if:
+     * <ul>
+     *     <li>Path is null</li>
+     *     <li>Path is not a "directory"</li>
+     *     <li>Path exists but cannot be read from</li>
+     * </ul>
+     *
+     * @param path      The path to lookup resources in.
+     * @param minDepth  Minimum folder depth from path to start searching (default: 0)
+     * @param maxDepth  Maximum folder depth from path to stop searching (default: null)
+     * @return empty set if nothing was found
+     * @throws IOException if there is an error searching the system.
+     */
+    List<Resource> search(String path, Integer minDepth, Integer maxDepth) throws IOException;
```

Comment:
Having a new search() version that is smarter allows us to add nice functionality like you have here.

However, just adding this new search function to the interface will break any extensions which simply implement this directly. I also see an unending set of other options people will want to add to the search logic over time which will lead to a large number of methods that ResourceAccessor implementations will have to think about and callers have to decide between.

I'm thinking it would be better to have ResourceAccessor.SearchOptions inner class with fields of minDepth, maxDepth, and recursive and change this new method to be `search(String path, SearchOptions searchOptions)` instead. That way future new like "file name pattern" or whatever can be included without having to make new method permutations.

Then, I'm wondering if we could default that new search method to call the existing `search(path, boolean)` call and do in-memory filtering of the results? That would allow ResourceAccessor implementors to generally just have the one method to implement and keep getting the additional functionality as it's added? But still leaving them to implement the new `search(String, SearchOptions)` if they feel they can do a better job?

Reply:
I really like that idea, the more extensible the better! Especially in this situation where things are community driven and there is this extension capability. I do like the default implementation in the interface to allow it to move forward without breaking current implementations and allow future implementors to override only if they want.

I can definitely take a stab at that... however I am not a Java programmer. I picked this up as we are using Liquibase for our project and this is the first time I've touched Java in ~15 years (and I've only been using Liquibase for ~ 1 month). As such, please feel free to let me know where improvements can be made as my Java knowledge is very outdated and I'm still learning Liquibase!